File: src/types/safety.ts
import type { Location } from './location';
import { User } from './user';

export type IncidentType =
  | 'harassment'
  | 'inappropriate'
  | 'impersonation'
  | 'scam'
  | 'emergency'
  | 'safety_check'
  | 'other';

export type IncidentStatus = 'pending' | 'resolved' | 'dismissed';

export type VerificationStatus =
  | 'unverified'
  | 'pending'
  | 'verified'
  | 'rejected';

export type SafetyAlertType =
  | 'sos'
  | 'meetup'
  | 'location'
  | 'custom'
  | 'emergency';
export type SafetyAlertStatus = 'pending' | 'active' | 'resolved' | 'dismissed';
export type SafetyCheckStatus = 'pending' | 'safe' | 'unsafe' | 'missed';

export interface EmergencyContactNotifyOn {
  sosAlert: boolean;
  meetupStart: boolean;
  meetupEnd: boolean;
  lowBattery: boolean;
  enterPrivacyZone: boolean;
  exitPrivacyZone: boolean;
}

export interface EmergencyContactFormData {
  firstName: string;
  lastName: string;
  phoneNumber?: string;
  email?: string;
  notifyOn: EmergencyContactNotifyOn;
}

export interface EmergencyContact {
  id: string;
  userId: string;
  firstName: string;
  lastName: string;
  phoneNumber?: string;
  email?: string;
  createdAt: string;
  updatedAt: string;
}

export interface EmergencyContactInput {
  firstName: string;
  lastName: string;
  phoneNumber?: string;
  email?: string;
}

// Old emergency contact type kept for reference
export interface EmergencyContactLegacy {
  id: string;
  userId: string;
  name: string;
  relationship: string;
  phoneNumber: string;
  email?: string;
  notifyOn: {
    sosAlert: boolean;
    meetupStart: boolean;
    meetupEnd: boolean;
    lowBattery: boolean;
    enterPrivacyZone: boolean;
    exitPrivacyZone: boolean;
  };
  createdAt: string;
  updatedAt: string;
}

export interface SafetyAlert {
  id: string;
  userId: string;
  type: SafetyAlertType;
  status: SafetyAlertStatus;
  location?: Location;
  message?: string;
  contactedEmergencyServices: boolean;
  notifiedContacts: string[];
  createdAt: string;
  updatedAt: string;
}

export interface SafetyCheckInput {
  type: 'meetup' | 'location' | 'custom';
  description: string;
  scheduledFor: string;
  location?: Location;
}

export interface SafetyCheck {
  id: string;
  userId: string;
  status: SafetyCheckStatus;
  scheduledTime: string;
  location?: Location;
  notes?: string;
  notifiedContacts: string[];
  createdAt: string;
  updatedAt: string;
  completedAt?: string;
}

export interface SafetyReport {
  id: string;
  reporterId: string;
  reportedUserId: string;
  type: 'harassment' | 'inappropriate' | 'spam' | 'scam' | 'other';
  description: string;
  evidence?: Array<{
    type: string;
    url: string;
  }>;
  status: 'pending' | 'reviewing' | 'resolved' | 'dismissed';
  adminNotes?: string;
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
}

export interface SafetyReportFormProps {
  reportedUserId: string;
  onSubmit: (report: Partial<SafetyReportNew>) => Promise<void>;
  onCancel: () => void;
}

export interface SafetyFeature {
  id: string;
  name: string;
  description: string;
  icon: string;
  enabled: boolean;
  requiresPermission?: boolean;
  permissions?: string[];
}

export interface SafetySettings {
  sosAlertEnabled: boolean;
  emergencyContacts: string[];
}

export interface SafetyReportOld {
  id: string;
  reporterId: string;
  reportedId: string;
  meetingId?: string;
  type: IncidentType;
  description: string;
  evidence: Evidence[];
  status: IncidentStatus;
  resolution?: string;
  location?: {
    latitude: number;
    longitude: number;
    accuracy: number;
  };
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
}

export interface Evidence {
  id: string;
  reportId: string;
  type: 'image' | 'video' | 'audio' | 'document' | 'other';
  url: string;
  thumbnail?: string;
  mimeType: string;
  size: number;
  createdAt: string;
}

export interface SafetyAlertOld {
  id: string;
  userId: string;
  type: 'emergency' | 'warning' | 'info';
  message: string;
  createdAt: Date;
  resolvedAt?: Date;
  location?: {
    latitude: number;
    longitude: number;
  };
}

export interface SafetyCheckOld {
  id: string;
  userId: string;
  type: 'meetup' | 'checkin' | 'custom';
  status: 'pending' | 'completed' | 'missed';
  scheduledFor: Date;
  completedAt?: Date;
  location?: {
    latitude: number;
    longitude: number;
  };
  notes?: string;
}

export interface UserBlock {
  id: string;
  blockerId: string;
  blockedId: string;
  reason?: string;
  createdAt: string;
  expiresAt?: string;
}

export interface EmergencyAlert {
  id: string;
  userId: string;
  type: 'sos' | 'meetup' | 'safety_check';
  location?: {
    latitude: number;
    longitude: number;
    accuracy: number;
  };
  status: IncidentStatus;
  message?: string;
  contactedEmergencyServices: boolean;
  notifiedContacts: string[];
  createdAt: string;
  resolvedAt?: string;
}

export interface SafetySettingsOld {
  emergencyContacts: EmergencyContactLegacy[];
  autoShareLocation: boolean;
  meetupCheckins: boolean;
  sosAlertEnabled: boolean;
  requireVerifiedMatch: boolean;
  blockList: UserBlock[];
  verificationStatus: VerificationStatus;
  verifiedAt?: string;
  lastSafetyCheck?: string;
}

export interface SafetySettingsNew {
  autoShareLocation: boolean;
  meetupCheckins: boolean;
  sosAlertEnabled: boolean;
  requireVerifiedMatch: boolean;
  emergencyContacts: EmergencyContact[];
}

export interface EmergencyContactNew {
  id: string;
  name: string;
  phone: string;
  email?: string;
  relationship: string;
  isVerified: boolean;
  userId: string;
  createdAt: string;
  updatedAt: string;
}

export interface SafetyAlertNew {
  id: string;
  userId: string;
  user?: User;
  type: SafetyAlertType;
  status: 'active' | 'resolved' | 'dismissed';
  location: {
    latitude: number;
    longitude: number;
    accuracy?: number;
    timestamp: Date;
  };
  message?: string;
  contacts?: string[];
  description?: string;
  evidence?: SafetyEvidence[];
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
}

export interface SafetyEvidence {
  id: string;
  alertId: string;
  type: 'image' | 'video' | 'audio' | 'text';
  url: string;
  description?: string;
  createdAt: string;
}

export interface SafetyCheckNew {
  id: string;
  userId: string;
  type: 'meetup' | 'location' | 'custom';
  status: 'pending' | 'completed' | 'missed';
  scheduledFor: string;
  completedAt?: string;
  location?: Location;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

export interface SafetyReportNew {
  id: string;
  reporterId: string;
  reportedUserId: string;
  type: 'harassment' | 'inappropriate' | 'spam' | 'scam' | 'other';
  description: string;
  evidence?: SafetyEvidence[];
  status: 'pending' | 'reviewing' | 'resolved' | 'dismissed';
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
}

export interface SafetyAlertContextType {
  alerts: SafetyAlertNew[];
  loading: boolean;
  error: string | null;
  addAlert: (alert: Partial<SafetyAlertNew>) => Promise<void>;
  dismissAlert: (alertId: string) => Promise<void>;
  resolveAlert: (alertId: string) => Promise<void>;
}

export interface SafetyCenterProps {
  userId: string;
  safetySettings: SafetySettings;
  onSettingsChange?: (settings: Partial<SafetySettingsNew>) => void;
}

export interface SafetyCheckListProps {
  checks: SafetyCheckNew[];
  onCheckComplete?: (checkId: string) => void;
}

export interface SafetyFeaturesProps {
  user: User;
  settings: SafetySettingsNew;
  onSettingsChange: (settings: Partial<SafetySettingsNew>) => void;
}

export interface EmergencyAlertProps {
  userId: string;
  onAlertTriggered?: (alert: SafetyAlertNew) => void;
}

export interface SafetyReportFormProps {
  reportedUserId: string;
  onSubmit: (report: Partial<SafetyReportNew>) => Promise<void>;
  onCancel: () => void;
}
export type SafetyAlertType = 'emergency' | 'help' | 'location';

export interface SafetyAlertNew {
  id?: string;
  type: SafetyAlertType;
  message: string | null;
  description: string | null;
  priority: string;
  location: {
    latitude: number;
    longitude: number;
    accuracy: number | null;
    timestamp: string;
  };
  dismissed: boolean;
  resolved: boolean;
  updatedAt?: Date;
  dismissedAt?: Date | null;
  resolvedAt?: Date | null;
}

export interface EmergencyContact {
  id: string;
  name: string;
  phoneNumber: string;
  email: string;
}

export interface SafetySettings {
  emergencyContacts: EmergencyContact[];
  autoShareLocation?: boolean;
  meetupCheckins?: boolean;
  sosAlertEnabled: boolean;
  requireVerifiedMatch?: boolean;
}

File: services/core-api/internal/models/safety.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type IncidentType string

const (
	IncidentTypeHarassment     IncidentType = "harassment"
	IncidentTypeInappropriate  IncidentType = "inappropriate"
	IncidentTypeImpersonation  IncidentType = "impersonation"
	IncidentTypeScam          IncidentType = "scam"
	IncidentTypeEmergency     IncidentType = "emergency"
	IncidentTypeOther         IncidentType = "other"
)

type IncidentStatus string

const (
	IncidentStatusPending    IncidentStatus = "pending"
	IncidentStatusReviewing  IncidentStatus = "reviewing"
	IncidentStatusResolved   IncidentStatus = "resolved"
	IncidentStatusDismissed  IncidentStatus = "dismissed"
)

// SafetyReport represents a user-submitted incident report
type SafetyReport struct {
	ID          uuid.UUID      `json:"id" gorm:"primaryKey;type:uuid"`
	ReporterID  uuid.UUID      `json:"reporter_id" gorm:"type:uuid;not null"`
	ReportedID  uuid.UUID      `json:"reported_id" gorm:"type:uuid;not null"`
	Type        IncidentType   `json:"type" gorm:"not null"`
	Description string         `json:"description"`
	Evidence    []Evidence     `json:"evidence" gorm:"foreignKey:ReportID"`
	Status      IncidentStatus `json:"status" gorm:"not null;default:'pending'"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	ResolvedAt  *time.Time     `json:"resolved_at"`
}

// Evidence represents supporting evidence for a safety report
type Evidence struct {
	ID        uuid.UUID `json:"id" gorm:"primaryKey;type:uuid"`
	ReportID  uuid.UUID `json:"report_id" gorm:"type:uuid;not null"`
	Type      string    `json:"type" gorm:"not null"` // e.g., "image", "chat_log", "screenshot"
	URL       string    `json:"url" gorm:"not null"`
	CreatedAt time.Time `json:"created_at"`
}

// EmergencyContact represents a user's emergency contact
type EmergencyContact struct {
	ID        uuid.UUID `json:"id" gorm:"primaryKey;type:uuid"`
	UserID    uuid.UUID `json:"user_id" gorm:"type:uuid;not null"`
	Name      string    `json:"name" gorm:"not null"`
	Phone     string    `json:"phone" gorm:"not null"`
	Email     string    `json:"email"`
	Relation  string    `json:"relation"`
	NotifyOn  []string  `json:"notify_on" gorm:"type:text[]"` // Array of incident types that trigger notification
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// UserBlock represents a user blocking another user
type UserBlock struct {
	ID          uuid.UUID `json:"id" gorm:"primaryKey;type:uuid"`
	BlockerID   uuid.UUID `json:"blocker_id" gorm:"type:uuid;not null"`
	BlockedID   uuid.UUID `json:"blocked_id" gorm:"type:uuid;not null"`
	Reason      string    `json:"reason"`
	CreatedAt   time.Time `json:"created_at"`
	ExpiresAt   *time.Time `json:"expires_at"` // Optional expiration for temporary blocks
}

// EmergencyAlert represents a real-time emergency alert
type EmergencyAlert struct {
	ID        uuid.UUID `json:"id" gorm:"primaryKey;type:uuid"`
	UserID    uuid.UUID `json:"user_id" gorm:"type:uuid;not null"`
	Type      string    `json:"type" gorm:"not null"`
	Location  *Location `json:"location" gorm:"embedded"`
	Status    string    `json:"status" gorm:"not null;default:'active'"`
	Message   string    `json:"message"`
	CreatedAt time.Time `json:"created_at"`
	ResolvedAt *time.Time `json:"resolved_at"`
}

// Location embedded type for EmergencyAlert
type Location struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Accuracy  float32 `json:"accuracy"`
}
